### DISCUSSION TITLE
Slow query parsing, not sure why

### DISCUSSION THREAD

**Discussion by emiasims** [tree-sitter/tree-sitter#1853] [Q&A] [answered]
I wasn't sure if this should be an issue, but it's not super actionable/specific, so here we are.

I wrote the parser at https://github.com/milisims/tree-sitter-org, which parses org files. The relevant bits to the story here are just that normal/non-structured text is parsed as a `(paragraph)` node, and on the main branch within the paragraph are `(expr)` nodes. These nodes are whitespace delimited and match segments of strings, numbers, and symbols.

In the **flatten-expr** branch of this parser, I'm removing `(expr)` nodes in favor of `(str)`, `(sym)`, and `(num)` nodes explicitly. The structure of the parsing is mostly very similar, previously they were anonymous nodes, and now they are named, and the parent `(expr)` is anonymous.

For example, a file containing "hi1!" previously parsed to `(paragraph (expr "str" "num" "sym"))`, and now parses to `(paragraph (str) (num) (sym))`

These changes come with a weird side effect, if I have a query that looks like:
On main, `(paragraph (expr) (expr) (expr) (expr) (expr) (expr) (expr) (expr) (expr) (expr))` (10 exprs), this compiles very quickly, less than 0.1s
On the new branch, the replacing (expr) with (str), this compiles very much more slowly, about 1 second.

The slowdown is very relevant, because I want to write complex queries to match objects dynamically.

To reproduce this,
``` bash
git clone git@github.com:milisims/tree-sitter-org.git
cd tree-sitter-org
git checkout flatten-expr  # to have npm install recent tree-sitter version, instead of 19.5
npm install
```

Then execute the following script, it just makes a query (paragraph (expr) ...) with `n` (expr)s, first on main, then on the other branch with `(str)`
``` bash
#!/bin/bash
set -euf -o pipefail

echo 'hi' > test.org
git checkout main
./node_modules/.bin/tree-sitter parse test.org > /dev/null
echo
echo "main branch with (expr)"
echo "query: '(paragraph (expr) ... )' with (expr) repeated n times"
for i in $(seq 2 30); do
  echo "(paragraph $(printf ' (expr) %.0s' $(seq $i)))" > q.scm
  echo -n "$i: "
  { time ./node_modules/.bin/tree-sitter query q.scm test.org; } 2>&1 | grep real | cut -f2 -dm
done
echo
echo
git checkout flatten-expr
./node_modules/.bin/tree-sitter parse test.org > /dev/null
echo
echo "query: '(paragraph (str) ... )' with (str) repeated n times"
for i in $(seq 2 20); do
  echo "(paragraph $(printf ' (str) %.0s' $(seq $i)))" > q.scm
  echo -n "$i: "
  { time ./node_modules/.bin/tree-sitter query q.scm test.org; } 2>&1 | grep real | cut -f2 -dm
done
rm q.scm test.org
```

The output on my machine looks like:
```
Switched to branch 'main'
Your branch is up to date with 'origin/main'.

main branch with (expr)
query: '(paragraph (expr) ... )' with (expr) repeated n times
2: 0.047s
3: 0.048s
4: 0.049s
5: 0.051s
6: 0.052s
7: 0.055s
8: 0.053s
9: 0.055s
10: 0.065s
11: 0.064s
12: 0.065s
13: 0.066s
14: 0.067s
15: 0.064s
16: 0.092s
17: 0.095s
18: 0.095s
19: 0.099s
20: 0.097s
21: 0.097s
22: 0.097s
23: 0.097s
24: 0.095s
25: 0.098s
26: 0.096s
27: 0.094s
28: 0.101s
29: 0.095s
30: 0.097s


Switched to branch 'flatten-expr'

query: '(paragraph (str) ... )' with (str) repeated n times
2: 0.049s
3: 0.069s
4: 0.109s
5: 0.122s
6: 0.138s
7: 0.145s
8: 0.154s
9: 0.162s
10: 0.894s
11: 0.984s
12: 1.049s
13: 1.122s
14: 1.187s
15: 1.253s
16: 1.316s
17: 1.430s
18: 1.420s
19: 1.431s
20: 1.462s
```

I went up to 30, but it stayed around 1.5s, but two of these expressions takes twice as long.

So I'm a bit at a loss. Any ideas?

Edit: I am testing this in neovim where I first observed the issue, I don't know if it's possible to test query parse time in the CLI directly, so I queried on a very simple file to minimize parse time. In neovim, you can set a similar test up quite easily with `vim._ts_parse_query` to get more direct parse times.

**Comment by amaanq** [comment, marked as answer]
This was fixed in #4614
